In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

# Setting display format to retina in matplotlib to see better quality images.
from IPython.display import set_matplotlib_formats
set_matplotlib_formats('retina')

# Lines below are just to ignore warnings
import warnings
warnings.filterwarnings('ignore')

from config import username, password
from sqlalchemy import create_engine
import psycopg2

In [2]:
# Dowanloading original msd dataset
msd_df = pd.read_csv('msd_summary.csv')
msd_df.head()

song_no  analysis_sample_rate                         audio_md5  \
0        1                 22050  aee9820911781c734e7694c5432990ca   
1        2                 22050  ed222d07c83bac7689d52753610a513a   
2        3                 22050  96c7104889a128fef84fa469d60e380c   
3        4                 22050  0f7da84b6b583e3846c7e022fb3a92a2   
4        5                 22050  228dd6392ad8001b0281f533f34c72fd   

   danceability   duration  end_of_fade_in  energy  idx_bars_confidence  \
0             0  252.05506           2.049       0                    0   
1             0  156.55138           0.258       0                    0   
2             0  138.97098           0.000       0                    0   
3             0  145.05751           0.000       0                    0   
4             0  514.29832           0.000       0                    0   

   idx_bars_start  idx_beats_confidence  ...  idx_artist_terms  \
0               0                     0  ...                 0   
1               0                     0  ...                 0   
2               0                     0  ...                 0   
3               0                     0  ...                 0   
4               0                     0  ...                 0   

   idx_similar_artists                               release  \
0                    0                 Monster Ballads X-Mas   
1                    0                           Karkuteillä   
2                    0                                Butter   
3                    0                               De Culo   
4                    0  Rene Ablaze Presents Winter Sessions   

   release_7digitalid  song_hotttnesss             song_id              title  \
0              633681         0.542899  SOQMMHC12AB0180CB8       Silent Night   
1              145266         0.299877  SOVFVAK12A8C1350D9        Tanssi vaan   
2              625706         0.617871  SOGTUKN12AB017F4F1  No One Could Ever   
3              199368              NaN  SOBNYVR12A8C13558C      Si Vos Querés   
4              209038              NaN  SOHSBXH12A8C13B0DF   Tangle Of Aspens   

   track_7digitalid  idx_artist_mbtags  year  
0           7032331                  0  2003  
1           1514808                  0  1995  
2           6945353                  0  2006  
3           2168257                  0  2003  
4           2264873                  0     0  

[5 rows x 54 columns]

In [3]:
# Basic information of msd_df
msd_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 54 columns):
 #   Column                          Non-Null Count    Dtype  
---  ------                          --------------    -----  
 0   song_no                         1000000 non-null  int64  
 1   analysis_sample_rate            1000000 non-null  int64  
 2   audio_md5                       1000000 non-null  object 
 3   danceability                    1000000 non-null  int64  
 4   duration                        1000000 non-null  float64
 5   end_of_fade_in                  1000000 non-null  float64
 6   energy                          1000000 non-null  int64  
 7   idx_bars_confidence             1000000 non-null  int64  
 8   idx_bars_start                  1000000 non-null  int64  
 9   idx_beats_confidence            1000000 non-null  int64  
 10  idx_beats_start                 1000000 non-null  int64  
 11  idx_sections_confidence         1000000 non-null  int64  
 12  i

In [4]:
# unique track count
print(f"Total number of unique track: {msd_df['track_id'].nunique()}")

Total number of unique track: 1000000


In [5]:
# unique artist count by artist_id
print(f"Total number of unique artists by artist_name: {msd_df['artist_id'].nunique()}")

Total number of unique artists by artist_name: 44745


In [6]:
# unique artist count by artist_name
print(f"Total number of unique artists by artist_name: {msd_df['artist_name'].nunique()}")

Total number of unique artists by artist_name: 72664


In [7]:
# unique artist_location count
print(f"Total number of unique locations of artist-origin: {msd_df['artist_location'].nunique()}")

Total number of unique locations of artist-origin: 5055


In [8]:
# unique album_of_release count by name
print(f"Total number of unique album of release by name: {msd_df['release'].nunique()}")

Total number of unique album of release by name: 149275


In [9]:
# unique album_of_release count by id
print(f"Total number of unique album of release by id: {msd_df['release_7digitalid'].nunique()}")

Total number of unique album of release by id: 177228


In [10]:
# Dowanloading the trimmed msd dataset
trimmed_msd = pd.read_csv('trimmed_msd.csv')
trimmed_msd.head()

song_no   duration  end_of_fade_in  key  key_confidence  loudness  mode  \
0        1  252.05506           2.049   10           0.777    -4.829     0   
1        2  156.55138           0.258    9           0.808   -10.555     1   
2        3  138.97098           0.000    7           0.418    -2.060     1   
3        4  145.05751           0.000    7           0.125    -4.654     1   
4        5  514.29832           0.000    5           0.097    -7.806     0   

   mode_confidence  start_of_fade_out    tempo  ...    artist_location  \
0            0.688            236.635   87.002  ...                NaN   
1            0.355            148.660  150.778  ...                NaN   
2            0.566            138.971  177.768  ...  Glasgow, Scotland   
3            0.451            138.687   87.433  ...                NaN   
4            0.290            506.717  140.035  ...                NaN   

            artist_id artist_7digitalid  artist_playmeid       artist_name  \
0  ARYZTJS1187B98C555              4069            44895  Faster Pussy cat   
1  ARMVN3U1187FB3A1EB            113480               -1  Karkkiautomaatti   
2  ARGEKB01187FB50750             63531               -1    Hudson Mohawke   
3  ARNWYLR1187B9B2F9C             65051            34000       Yerba Brava   
4  AREQDTE1269FB37231            158279               -1        Der Mystic   

                                release release_7digitalid  song_hotttnesss  \
0                 Monster Ballads X-Mas             633681         0.542899   
1                           Karkuteillä             145266         0.299877   
2                                Butter             625706         0.617871   
3                               De Culo             199368              NaN   
4  Rene Ablaze Presents Winter Sessions             209038              NaN   

               title  year  
0       Silent Night  2003  
1        Tanssi vaan  1995  
2  No One Could Ever  2006  
3      Si Vos Querés  2003  
4   Tangle Of Aspens     0  

[5 rows x 25 columns]

In [11]:
# Basic information of trimmed_msd_df
trimmed_msd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 25 columns):
 #   Column                     Non-Null Count    Dtype  
---  ------                     --------------    -----  
 0   song_no                    1000000 non-null  int64  
 1   duration                   1000000 non-null  float64
 2   end_of_fade_in             1000000 non-null  float64
 3   key                        1000000 non-null  int64  
 4   key_confidence             1000000 non-null  float64
 5   loudness                   1000000 non-null  float64
 6   mode                       1000000 non-null  int64  
 7   mode_confidence            1000000 non-null  float64
 8   start_of_fade_out          1000000 non-null  float64
 9   tempo                      1000000 non-null  float64
 10  time_signature             1000000 non-null  int64  
 11  time_signature_confidence  1000000 non-null  float64
 12  track_id                   1000000 non-null  object 
 13  artist_famili

In [12]:
# Change datatype for year to date
# trimmed_msd['year'] = pd.to_datetime(trimmed_msd['year'], format='%Y')

In [13]:
# check for missing data
missing_data = trimmed_msd.isnull().sum()
missing_data

song_no                           0
duration                          0
end_of_fade_in                    0
key                               0
key_confidence                    0
loudness                          0
mode                              0
mode_confidence                   0
start_of_fade_out                 0
tempo                             0
time_signature                    0
time_signature_confidence         0
track_id                          0
artist_familiarity              185
artist_hotttnesss                12
artist_location              487129
artist_id                         0
artist_7digitalid                 0
artist_playmeid                   0
artist_name                       0
release                           5
release_7digitalid                0
song_hotttnesss              418035
title                            15
year                              0
dtype: int64

In [14]:
# check for percentage of missing data
percent_missing = round((trimmed_msd.isnull().sum())/(trimmed_msd.isnull().count())*100,2)
percent_missing.sort_values(ascending=False).head(2)

artist_location    48.71
song_hotttnesss    41.80
dtype: float64

In [15]:
 # Drop artist_location column due to high percentage of missing data
trimmed_msd.drop(['artist_location'], axis=1, inplace=True)
trimmed_msd.head()

song_no   duration  end_of_fade_in  key  key_confidence  loudness  mode  \
0        1  252.05506           2.049   10           0.777    -4.829     0   
1        2  156.55138           0.258    9           0.808   -10.555     1   
2        3  138.97098           0.000    7           0.418    -2.060     1   
3        4  145.05751           0.000    7           0.125    -4.654     1   
4        5  514.29832           0.000    5           0.097    -7.806     0   

   mode_confidence  start_of_fade_out    tempo  ...  artist_hotttnesss  \
0            0.688            236.635   87.002  ...           0.394032   
1            0.355            148.660  150.778  ...           0.356992   
2            0.566            138.971  177.768  ...           0.437504   
3            0.451            138.687   87.433  ...           0.372349   
4            0.290            506.717  140.035  ...           0.000000   

            artist_id artist_7digitalid  artist_playmeid       artist_name  \
0  ARYZTJS1187B98C555              4069            44895  Faster Pussy cat   
1  ARMVN3U1187FB3A1EB            113480               -1  Karkkiautomaatti   
2  ARGEKB01187FB50750             63531               -1    Hudson Mohawke   
3  ARNWYLR1187B9B2F9C             65051            34000       Yerba Brava   
4  AREQDTE1269FB37231            158279               -1        Der Mystic   

                                release  release_7digitalid  song_hotttnesss  \
0                 Monster Ballads X-Mas              633681         0.542899   
1                           Karkuteillä              145266         0.299877   
2                                Butter              625706         0.617871   
3                               De Culo              199368              NaN   
4  Rene Ablaze Presents Winter Sessions              209038              NaN   

               title  year  
0       Silent Night  2003  
1        Tanssi vaan  1995  
2  No One Could Ever  2006  
3      Si Vos Querés  2003  
4   Tangle Of Aspens     0  

[5 rows x 24 columns]

In [16]:
# Drop the observations with missing data
trimmed_msd.dropna(inplace=True)
trimmed_msd.head()

song_no   duration  end_of_fade_in  key  key_confidence  loudness  mode  \
0         1  252.05506           2.049   10           0.777    -4.829     0   
1         2  156.55138           0.258    9           0.808   -10.555     1   
2         3  138.97098           0.000    7           0.418    -2.060     1   
12       13  301.60934           0.000    1           0.450    -4.882     1   
13       14  318.45832           0.502   10           0.764   -10.670     1   

    mode_confidence  start_of_fade_out    tempo  ...  artist_hotttnesss  \
0             0.688            236.635   87.002  ...           0.394032   
1             0.355            148.660  150.778  ...           0.356992   
2             0.566            138.971  177.768  ...           0.437504   
12            0.520            291.405   85.340  ...           0.451579   
13            0.627            306.265   67.567  ...           0.406985   

             artist_id artist_7digitalid  artist_playmeid       artist_name  \
0   ARYZTJS1187B98C555              4069            44895  Faster Pussy cat   
1   ARMVN3U1187FB3A1EB            113480               -1  Karkkiautomaatti   
2   ARGEKB01187FB50750             63531               -1    Hudson Mohawke   
12  ARVIT0V1187B9A7CDE            165382               -1      Danny Diablo   
13  AREMPER1187B9AEB42              5823               -1         Tiger Lou   

                             release  release_7digitalid  song_hotttnesss  \
0              Monster Ballads X-Mas              633681         0.542899   
1                        Karkuteillä              145266         0.299877   
2                             Butter              625706         0.617871   
12  International Hardcore Superstar              620618         0.392009   
13                         The Loyal              204414         0.463490   

                                  title  year  
0                          Silent Night  2003  
1                           Tanssi vaan  1995  
2                     No One Could Ever  2006  
12  Cold Beer feat. Prince Metropolitan     0  
13                               Pilots  2005  

[5 rows x 24 columns]

In [17]:
# Basic information of trimmed_msd_df
trimmed_msd.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 581909 entries, 0 to 999999
Data columns (total 24 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   song_no                    581909 non-null  int64  
 1   duration                   581909 non-null  float64
 2   end_of_fade_in             581909 non-null  float64
 3   key                        581909 non-null  int64  
 4   key_confidence             581909 non-null  float64
 5   loudness                   581909 non-null  float64
 6   mode                       581909 non-null  int64  
 7   mode_confidence            581909 non-null  float64
 8   start_of_fade_out          581909 non-null  float64
 9   tempo                      581909 non-null  float64
 10  time_signature             581909 non-null  int64  
 11  time_signature_confidence  581909 non-null  float64
 12  track_id                   581909 non-null  object 
 13  artist_familiarity         58

In [18]:
# unique track count for the trimmed dataset
print(f"Total number of unique track: {trimmed_msd['track_id'].nunique()}")

Total number of unique track: 581909


In [19]:
# unique artist count for the trimmed dataset (by artist_id)
print(f"Total number of unique artists by artist_id: {trimmed_msd['artist_id'].nunique()}")

Total number of unique artists by artist_id: 36087


In [20]:
# unique artist count for the trimmed dataset (by artist_7digitalid)
print(f"Total number of unique artists by artist_7digitalid: {trimmed_msd['artist_7digitalid'].nunique()}")

Total number of unique artists by artist_7digitalid: 35598


In [21]:
# unique artist count for the trimmed dataset (by artist_playmeid)
print(f"Total number of unique artists by artist_playmeid: {trimmed_msd['artist_playmeid'].nunique()}") 

Total number of unique artists by artist_playmeid: 12902


In [22]:
# unique artist count for the trimmed dataset (by artist_name)
print(f"Total number of unique artists by artist_name: {trimmed_msd['artist_name'].nunique()}")

Total number of unique artists by artist_name: 51751


In [23]:
# unique album_of_release count for the trimmed dataset (by name) 
print(f"Total number of unique album of release by name: {trimmed_msd['release'].nunique()}")

Total number of unique album of release by name: 108172


In [24]:
# unique companies_of_release count for the trimmed dataset (by id) 
print(f"Total number of unique album of release by name: {trimmed_msd['release_7digitalid'].nunique()}")

Total number of unique album of release by name: 125815


In [25]:
# Basic statistics for the clean dataset --- NOT YET. 
# Need further cleaning
trimmed_msd.describe()

song_no       duration  end_of_fade_in            key  \
count   581909.000000  581909.000000   581909.000000  581909.000000   
mean    500593.424517     247.472243        0.833324       5.316185   
std     288763.695996     118.622184        3.428424       3.592778   
min          1.000000       0.522000        0.000000       0.000000   
25%     250400.000000     183.300770        0.000000       2.000000   
50%     501264.000000     230.033830        0.194000       5.000000   
75%     750648.000000     286.170980        0.426000       9.000000   
max    1000000.000000    3032.580770      991.359000      11.000000   

       key_confidence       loudness          mode  mode_confidence  \
count   581909.000000  581909.000000  581909.00000    581909.000000   
mean         0.444725      -9.705346       0.66503         0.475587   
std          0.276527       4.967490       0.47198         0.192359   
min          0.000000     -55.751000       0.00000         0.000000   
25%          0.216000     -12.147000       0.00000         0.360000   
50%          0.464000      -8.594000       1.00000         0.486000   
75%          0.654000      -6.139000       1.00000         0.606000   
max          1.000000       4.318000       1.00000         1.000000   

       start_of_fade_out          tempo  time_signature  \
count      581909.000000  581909.000000   581909.000000   
mean          238.332774     124.694522        3.615234   
std           116.477629      34.969851        1.199780   
min             0.522000       0.000000        0.000000   
25%           175.520000      98.505000        3.000000   
50%           220.700000     122.069000        4.000000   
75%           275.638000     145.604000        4.000000   
max          3030.622000     284.208000        7.000000   

       time_signature_confidence  artist_familiarity  artist_hotttnesss  \
count              581909.000000       581909.000000      581909.000000   
mean                    0.519018            0.579859           0.398976   
std                     0.370371            0.130225           0.112635   
min                     0.000000            0.000000           0.000000   
25%                     0.138000            0.504933           0.352843   
50%                     0.560000            0.582700           0.399452   
75%                     0.870000            0.659129           0.456170   
max                     1.000000            1.000000           1.082503   

       artist_7digitalid  artist_playmeid  release_7digitalid  \
count      581909.000000    581909.000000       581909.000000   
mean       118581.057543     26480.363058       364752.068148   
std        140135.330716     47387.590191       236626.613808   
min            -1.000000        -1.000000           -1.000000   
25%         14355.000000        -1.000000       163978.000000   
50%         54572.000000       260.000000       323364.000000   
75%        176113.000000     34211.000000       560107.000000   
max        817066.000000    242965.000000       823606.000000   

       song_hotttnesss           year  
count    581909.000000  581909.000000  
mean          0.356080    1190.441535  
std           0.234431     981.452971  
min           0.000000       0.000000  
25%           0.215080       0.000000  
50%           0.377532    1992.000000  
75%           0.531985    2004.000000  
max           1.000000    2011.000000

In [26]:
# Check for outliers on target variable and key features ---- NOT CHECKED
# Finalize csv for clean data ---- NOT CHECKED
# Create another csv file: just for the artist_name and create artist_id ---- Checked
# Create another csv file: just for the release and create release_id ---- Checked

In [27]:
# Create a csv file for postgreSQL
trimmed_msd.drop(['song_no', 'artist_7digitalid', 'artist_playmeid', 'artist_playmeid', 'artist_id', 'release_7digitalid'], axis=1, inplace=True)
trimmed_msd

duration  end_of_fade_in  key  key_confidence  loudness  mode  \
0       252.05506           2.049   10           0.777    -4.829     0   
1       156.55138           0.258    9           0.808   -10.555     1   
2       138.97098           0.000    7           0.418    -2.060     1   
12      301.60934           0.000    1           0.450    -4.882     1   
13      318.45832           0.502   10           0.764   -10.670     1   
...           ...             ...  ...             ...       ...   ...   
999990  228.41424           0.101   11           0.800    -4.843     0   
999991  258.76853           0.287    4           0.247    -9.188     0   
999992  241.21424           1.007    0           0.502    -5.303     1   
999994  185.33832           0.000    1           0.040    -5.968     0   
999999  295.07873           0.000    6           0.634    -5.801     1   

        mode_confidence  start_of_fade_out    tempo  time_signature  \
0                 0.688            236.635   87.002               4   
1                 0.355            148.660  150.778               1   
2                 0.566            138.971  177.768               4   
12                0.520            291.405   85.340               5   
13                0.627            306.265   67.567               3   
...                 ...                ...      ...             ...   
999990            0.690            208.451  130.155               4   
999991            0.268            251.315   80.364               3   
999992            0.457            238.539  148.354               4   
999994            0.412            170.974  120.009               4   
999999            0.445            292.182  164.037               4   

        time_signature_confidence            track_id  artist_familiarity  \
0                           0.940  TRMMMYQ128F932D901            0.649822   
1                           0.000  TRMMMKD128F425225D            0.439604   
2                           0.446  TRMMMRX128F93187D9            0.643681   
12                          0.581  TRMMMCJ128F930BFF8            0.622005   
13                          0.676  TRMMMBW128F4260CAE            0.621826   
...                           ...                 ...                 ...   
999990                      0.332  TRYYYYS12903D0605A            0.735147   
999991                      0.493  TRYYYLF128E0789A44            0.586555   
999992                      0.508  TRYYYZM128F428E804            0.592700   
999994                      0.734  TRYYYNQ128F92E0292            0.699599   
999999                      0.672  TRYYYVU12903CD01E3            0.552977   

        artist_hotttnesss                                        artist_name  \
0                0.394032                                   Faster Pussy cat   
1                0.356992                                   Karkkiautomaatti   
2                0.437504                                     Hudson Mohawke   
12               0.451579                                       Danny Diablo   
13               0.406985                                          Tiger Lou   
...                   ...                                                ...   
999990           0.471100               Taylor Hawkins & The Coattail Riders   
999991           0.340153                                          Bah Samba   
999992           0.429545                                            SKYCLAD   
999994           0.392061  Killer Mike feat. Gangsta Pill and Nario of Gr...   
999999           0.454869                                              Texta   

                                                  release  song_hotttnesss  \
0                                   Monster Ballads X-Mas         0.542899   
1                                             Karkuteillä         0.299877   
2                                                  Butter         0.617871   
12                       International Hardcore Superstar         0.3

In [28]:
# Create a csv file for postgreSQL
postgres_msd = trimmed_msd
postgres_msd

duration  end_of_fade_in  key  key_confidence  loudness  mode  \
0       252.05506           2.049   10           0.777    -4.829     0   
1       156.55138           0.258    9           0.808   -10.555     1   
2       138.97098           0.000    7           0.418    -2.060     1   
12      301.60934           0.000    1           0.450    -4.882     1   
13      318.45832           0.502   10           0.764   -10.670     1   
...           ...             ...  ...             ...       ...   ...   
999990  228.41424           0.101   11           0.800    -4.843     0   
999991  258.76853           0.287    4           0.247    -9.188     0   
999992  241.21424           1.007    0           0.502    -5.303     1   
999994  185.33832           0.000    1           0.040    -5.968     0   
999999  295.07873           0.000    6           0.634    -5.801     1   

        mode_confidence  start_of_fade_out    tempo  time_signature  \
0                 0.688            236.635   87.002               4   
1                 0.355            148.660  150.778               1   
2                 0.566            138.971  177.768               4   
12                0.520            291.405   85.340               5   
13                0.627            306.265   67.567               3   
...                 ...                ...      ...             ...   
999990            0.690            208.451  130.155               4   
999991            0.268            251.315   80.364               3   
999992            0.457            238.539  148.354               4   
999994            0.412            170.974  120.009               4   
999999            0.445            292.182  164.037               4   

        time_signature_confidence            track_id  artist_familiarity  \
0                           0.940  TRMMMYQ128F932D901            0.649822   
1                           0.000  TRMMMKD128F425225D            0.439604   
2                           0.446  TRMMMRX128F93187D9            0.643681   
12                          0.581  TRMMMCJ128F930BFF8            0.622005   
13                          0.676  TRMMMBW128F4260CAE            0.621826   
...                           ...                 ...                 ...   
999990                      0.332  TRYYYYS12903D0605A            0.735147   
999991                      0.493  TRYYYLF128E0789A44            0.586555   
999992                      0.508  TRYYYZM128F428E804            0.592700   
999994                      0.734  TRYYYNQ128F92E0292            0.699599   
999999                      0.672  TRYYYVU12903CD01E3            0.552977   

        artist_hotttnesss                                        artist_name  \
0                0.394032                                   Faster Pussy cat   
1                0.356992                                   Karkkiautomaatti   
2                0.437504                                     Hudson Mohawke   
12               0.451579                                       Danny Diablo   
13               0.406985                                          Tiger Lou   
...                   ...                                                ...   
999990           0.471100               Taylor Hawkins & The Coattail Riders   
999991           0.340153                                          Bah Samba   
999992           0.429545                                            SKYCLAD   
999994           0.392061  Killer Mike feat. Gangsta Pill and Nario of Gr...   
999999           0.454869                                              Texta   

                                                  release  song_hotttnesss  \
0                                   Monster Ballads X-Mas         0.542899   
1                                             Karkuteillä         0.299877   
2                                                  Butter         0.617871   
12                       International Hardcore Superstar         0.3

In [29]:
# Create a csv file for postgreSQL
postgres_msd.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 581909 entries, 0 to 999999
Data columns (total 19 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   duration                   581909 non-null  float64
 1   end_of_fade_in             581909 non-null  float64
 2   key                        581909 non-null  int64  
 3   key_confidence             581909 non-null  float64
 4   loudness                   581909 non-null  float64
 5   mode                       581909 non-null  int64  
 6   mode_confidence            581909 non-null  float64
 7   start_of_fade_out          581909 non-null  float64
 8   tempo                      581909 non-null  float64
 9   time_signature             581909 non-null  int64  
 10  time_signature_confidence  581909 non-null  float64
 11  track_id                   581909 non-null  object 
 12  artist_familiarity         581909 non-null  float64
 13  artist_hotttnesss          58

In [30]:
# Create a csv file for postgreSQL
postgres_msd.to_csv('postgres_msd.csv', index=False)

In [31]:
# Create a csv file: just for the artist_name with artist_newID
artist = trimmed_msd[['artist_name']].drop_duplicates()
artist

artist_name
0                                        Faster Pussy cat
1                                        Karkkiautomaatti
2                                          Hudson Mohawke
12                                           Danny Diablo
13                                              Tiger Lou
...                                                   ...
999889  Ben Harney / Jennifer Holliday / Sheryl Lee Ra...
999916                                      Kim Hyun-Chul
999955                                    Pancho's Lament
999961  Ashley Tisdale/High School Musical Cast/Jemma ...
999994  Killer Mike feat. Gangsta Pill and Nario of Gr...

[51751 rows x 1 columns]

In [32]:
# Create a csv file: just for the artist_name with artist_newID
artist['artist_newid'] = artist.reset_index().index + 1
artist

artist_name  artist_newid
0                                        Faster Pussy cat             1
1                                        Karkkiautomaatti             2
2                                          Hudson Mohawke             3
12                                           Danny Diablo             4
13                                              Tiger Lou             5
...                                                   ...           ...
999889  Ben Harney / Jennifer Holliday / Sheryl Lee Ra...         51747
999916                                      Kim Hyun-Chul         51748
999955                                    Pancho's Lament         51749
999961  Ashley Tisdale/High School Musical Cast/Jemma ...         51750
999994  Killer Mike feat. Gangsta Pill and Nario of Gr...         51751

[51751 rows x 2 columns]

In [33]:
# Create a csv file: just for the artist_name with artist_newID
artist.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 51751 entries, 0 to 999994
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   artist_name   51751 non-null  object
 1   artist_newid  51751 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 1.2+ MB


In [34]:
# Create artist.csv file with unique artist_name and artist_newID
artist.to_csv('artist.csv', index=False)

In [35]:
# Create a csv file: just for release with release_newID
release = trimmed_msd[['release']].drop_duplicates()
release

release
0                  Monster Ballads X-Mas
1                            Karkuteillä
2                                 Butter
12      International Hardcore Superstar
13                             The Loyal
...                                  ...
999903               Dirty Laugh Remixes
999916            Kid's Pop 'Love Is...'
999926               Il Éait Une Fois...
999948                      Nachtexpress
999990                          Way Down

[108172 rows x 1 columns]

In [36]:
# Create a csv file: just for release with release_newID
release['release_newid'] = release.reset_index().index + 1
release

release  release_newid
0                  Monster Ballads X-Mas              1
1                            Karkuteillä              2
2                                 Butter              3
12      International Hardcore Superstar              4
13                             The Loyal              5
...                                  ...            ...
999903               Dirty Laugh Remixes         108168
999916            Kid's Pop 'Love Is...'         108169
999926               Il Éait Une Fois...         108170
999948                      Nachtexpress         108171
999990                          Way Down         108172

[108172 rows x 2 columns]

In [37]:
# Create a csv file: just for release with release_newID
release.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 108172 entries, 0 to 999990
Data columns (total 2 columns):
 #   Column         Non-Null Count   Dtype 
---  ------         --------------   ----- 
 0   release        108172 non-null  object
 1   release_newid  108172 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 2.5+ MB


In [38]:
# Create release.csv file with unique release and release_newID
release.to_csv('release.csv', index=False)

### Connection to Postgres & SQL Query

In [39]:
# Create Connection to Database
engine =create_engine(f'postgresql://{username}:{password}@localhost:5432/msd_DB')
connection= engine.connect()

In [40]:
# Exporting artist table to postgres database; For Large Datasets use Chucksize
artist.to_sql('artist', con=connection, if_exists='replace', index=False, chunksize = 100 )

In [41]:
# Exporting release table to postgres database; For Large Datasets use Chucksize
release.to_sql('release', con=connection, if_exists='replace', index=False, chunksize = 100 )

In [42]:
# Exporting release table to postgres database; For Large Datasets use Chucksize
postgres_msd.to_sql('postgres_msd', con=connection, if_exists='replace', index=False, chunksize = 100 )

In [43]:
final_msd = pd.read_sql(
'select b.track_id, b.title, a.artist_newid, a.artist_name, b.artist_familiarity, b.artist_hotttnesss,\
c.release_newid, c.release, b.year, b.duration, b.end_of_fade_in, b.start_of_fade_out, b.tempo, b.loudness,\
b.key, b.key_confidence, b.mode, b.mode_confidence, b.time_signature, b.time_signature_confidence, b.song_hotttnesss \
from artist a join postgres_msd b \
on a.artist_name = b.artist_name \
join release c on b.release = c.release \
order by a.artist_newid',
connection)

final_msd.head(50)

track_id                                              title  \
0   TRMMMYQ128F932D901                                       Silent Night   
1   TRNSZOC128F425224E                                        Hei Johanna   
2   TRABGDV128F4252252                                Toivon että huomaat   
3   TRMMMKD128F425225D                                        Tanssi vaan   
4   TRGBNVG128F425224D                                      Jää beibi jää   
5   TRQGNIX128F425224F                              Aina vaan jaa jaa jaa   
6   TRWUBYW128F4252258                                          Äl-oo-vee   
7   TRBZRME128F425225E                                    Takaisin en tuu   
8   TRFOSNY128F4252255                                Paina kaasua_ Honey   
9   TRQRDVN128F425225F                                   Annathan anteeks   
10  TRFXTTP128F4252259                               Ei oo ei oo toivokaa   
11  TRMLAVN128F4252261                                    Yeah yeah Jenni   
12  TRSJKBK128F4252256                   En kai koskaan löydä sitä oikeaa   
13  TRQJXBW128F425225C                                    Taaskin turhaan   
14  TRULYSN128F425225A                            Särkyneen sydämen twist   
15  TRVVNZU128F4252251                       Kai vielä joskus muistat mua   
16  TRKWTAK128F4252253                                   Hei okei mä meen   
17  TRWAUOR128F9318794                                      Acoustic Lady   
18  TRDQCHT128F931884B                            Allhot (feat. Nadsroic)   
19  TRGFBFN128F931877C                                           ZOo00OOm   
20  TRMZFDE128F93187A5                                           Rising 5   
21  TRFDRLT128F93187BF               Just Decided (feat. Olivier Daysoul)   
22  TRTTLBE128F9318835                                      Star Crackout   
23  TRMMMRX128F93187D9                                  No One Could Ever   
24  TRGMPUJ128F931871E                                          Gluetooth   
25  TRXLDMD128F93187E9                                        Velvet Peel   
26  TRPHCUJ128F931875F                                        Fruit Touch   
27  TRZFTDP128F9318744                                                3.3   
28  TRXBJIG128F9318822                                               FUSE   
29  TRWJVDX128EF356E90                                      Riding Chrome   
30  TRJECRX128F9318866                                        Black n Red   
31  TRVLPQX128F9318707                                      Shower Melody   
32  TREMHLW128F931872E              Joy Fantastic (feat. Olivier Daysoul)   
33  TRVURZH128F93187B8                                     Twistclip Loop   
34  TRZLVHB128F9318753                                              Trykk   
35  TRHGWQQ128F930C00D                  Goodbye Farewell feat. Toby Morse   
36  TRUQSRP128F930BFFD                                     Shot Gun Blast   
37  TRKZVUF128F930BFF9                  Give Me A Chance feat. White Lion   
38  TRECGMA128F930BFF6          Working Class featuring Natasha Nicholson   
39  TRHUTVG128F930BFFB                                        Blood Money   
40  TRKLBVH128F429C78E                                         Banged Out   
41  TRLYFVN128F429C796  Satanic Shamrocks (Features Big left_ Danny Bo...   
42  TRBLMUJ128F933F36D                                              Outro   
43  TRYSBRP128F933F32F                             Intro / Blow Your Mind   
44  TRHFAIK128F933F371                                       Private Hell   
45  TRECEMQ128F429C79A                                    The Harsh Truth   
46  TROYWEU128F930BFF2                                       Kill Em Dead   
47  TREJOQD128F930C003  The Chosen Few feat. Sick Jacken_ White Trash ...   
48  TRRFBRO128F930C00A  Queens Killaz feat. Prince Metropolitan and Cynic   
49  TRVJQZX128F930BFEF  Sex And Violence feat. Everlast And Tim Armstrong   

    artist_newid       artist_name  artist_familiarity  artist_hotttnesss  \
0        

In [44]:
final_msd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 581909 entries, 0 to 581908
Data columns (total 21 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   track_id                   581909 non-null  object 
 1   title                      581909 non-null  object 
 2   artist_newid               581909 non-null  int64  
 3   artist_name                581909 non-null  object 
 4   artist_familiarity         581909 non-null  float64
 5   artist_hotttnesss          581909 non-null  float64
 6   release_newid              581909 non-null  int64  
 7   release                    581909 non-null  object 
 8   year                       581909 non-null  int64  
 9   duration                   581909 non-null  float64
 10  end_of_fade_in             581909 non-null  float64
 11  start_of_fade_out          581909 non-null  float64
 12  tempo                      581909 non-null  float64
 13  loudness                   58

### EDA for ML Model

In [45]:
# Creating a new feature called song_popularity based on whether song_hotness is above or below mean 
song_hotness_mean = trimmed_msd['song_hotttnesss'].mean()
song_hotness_mean

0.356080330963952

In [46]:
# Creating a new feature called song_popularity based on whether song_hotness is above or below mean 
trimmed_msd['song_popularity'] = trimmed_msd['song_hotttnesss']
trimmed_msd.head()

duration  end_of_fade_in  key  key_confidence  loudness  mode  \
0   252.05506           2.049   10           0.777    -4.829     0   
1   156.55138           0.258    9           0.808   -10.555     1   
2   138.97098           0.000    7           0.418    -2.060     1   
12  301.60934           0.000    1           0.450    -4.882     1   
13  318.45832           0.502   10           0.764   -10.670     1   

    mode_confidence  start_of_fade_out    tempo  time_signature  \
0             0.688            236.635   87.002               4   
1             0.355            148.660  150.778               1   
2             0.566            138.971  177.768               4   
12            0.520            291.405   85.340               5   
13            0.627            306.265   67.567               3   

    time_signature_confidence            track_id  artist_familiarity  \
0                       0.940  TRMMMYQ128F932D901            0.649822   
1                       0.000  TRMMMKD128F425225D            0.439604   
2                       0.446  TRMMMRX128F93187D9            0.643681   
12                      0.581  TRMMMCJ128F930BFF8            0.622005   
13                      0.676  TRMMMBW128F4260CAE            0.621826   

    artist_hotttnesss       artist_name                           release  \
0            0.394032  Faster Pussy cat             Monster Ballads X-Mas   
1            0.356992  Karkkiautomaatti                       Karkuteillä   
2            0.437504    Hudson Mohawke                            Butter   
12           0.451579      Danny Diablo  International Hardcore Superstar   
13           0.406985         Tiger Lou                         The Loyal   

    song_hotttnesss                                title  year  \
0          0.542899                         Silent Night  2003   
1          0.299877                          Tanssi vaan  1995   
2          0.617871                    No One Could Ever  2006   
12         0.392009  Cold Beer feat. Prince Metropolitan     0   
13         0.463490                               Pilots  2005   

    song_popularity  
0          0.542899  
1          0.299877  
2          0.617871  
12         0.392009  
13         0.463490

In [47]:
# Creating a new feature called song_popularity based on whether song_hotness is above or below mean 
trimmed_msd['song_popularity'] = trimmed_msd['song_hotttnesss'].apply(lambda x: 1 if x > song_hotness_mean else 0)
trimmed_msd.head()

duration  end_of_fade_in  key  key_confidence  loudness  mode  \
0   252.05506           2.049   10           0.777    -4.829     0   
1   156.55138           0.258    9           0.808   -10.555     1   
2   138.97098           0.000    7           0.418    -2.060     1   
12  301.60934           0.000    1           0.450    -4.882     1   
13  318.45832           0.502   10           0.764   -10.670     1   

    mode_confidence  start_of_fade_out    tempo  time_signature  \
0             0.688            236.635   87.002               4   
1             0.355            148.660  150.778               1   
2             0.566            138.971  177.768               4   
12            0.520            291.405   85.340               5   
13            0.627            306.265   67.567               3   

    time_signature_confidence            track_id  artist_familiarity  \
0                       0.940  TRMMMYQ128F932D901            0.649822   
1                       0.000  TRMMMKD128F425225D            0.439604   
2                       0.446  TRMMMRX128F93187D9            0.643681   
12                      0.581  TRMMMCJ128F930BFF8            0.622005   
13                      0.676  TRMMMBW128F4260CAE            0.621826   

    artist_hotttnesss       artist_name                           release  \
0            0.394032  Faster Pussy cat             Monster Ballads X-Mas   
1            0.356992  Karkkiautomaatti                       Karkuteillä   
2            0.437504    Hudson Mohawke                            Butter   
12           0.451579      Danny Diablo  International Hardcore Superstar   
13           0.406985         Tiger Lou                         The Loyal   

    song_hotttnesss                                title  year  \
0          0.542899                         Silent Night  2003   
1          0.299877                          Tanssi vaan  1995   
2          0.617871                    No One Could Ever  2006   
12         0.392009  Cold Beer feat. Prince Metropolitan     0   
13         0.463490                               Pilots  2005   

    song_popularity  
0                 1  
1                 0  
2                 1  
12                1  
13                1

In [48]:
trimmed_msd['song_popularity'].value_counts()

1    311031
0    270878
Name: song_popularity, dtype: int64

In [49]:
X = trimmed_msd[['duration', 'end_of_fade_in', 'key', 'loudness', 'mode', 'start_of_fade_out', 'tempo', 'time_signature', 'artist_familiarity', 'artist_hotttnesss', 'year']]
y = trimmed_msd['song_popularity']

In [50]:
X

duration  end_of_fade_in  key  loudness  mode  start_of_fade_out  \
0       252.05506           2.049   10    -4.829     0            236.635   
1       156.55138           0.258    9   -10.555     1            148.660   
2       138.97098           0.000    7    -2.060     1            138.971   
12      301.60934           0.000    1    -4.882     1            291.405   
13      318.45832           0.502   10   -10.670     1            306.265   
...           ...             ...  ...       ...   ...                ...   
999990  228.41424           0.101   11    -4.843     0            208.451   
999991  258.76853           0.287    4    -9.188     0            251.315   
999992  241.21424           1.007    0    -5.303     1            238.539   
999994  185.33832           0.000    1    -5.968     0            170.974   
999999  295.07873           0.000    6    -5.801     1            292.182   

          tempo  time_signature  artist_familiarity  artist_hotttnesss  year  
0        87.002               4            0.649822           0.394032  2003  
1       150.778               1            0.439604           0.356992  1995  
2       177.768               4            0.643681           0.437504  2006  
12       85.340               5            0.622005           0.451579     0  
13       67.567               3            0.621826           0.406985  2005  
...         ...             ...                 ...                ...   ...  
999990  130.155               4            0.735147           0.471100  2010  
999991   80.364               3            0.586555           0.340153  2002  
999992  148.354               4            0.592700           0.429545  1996  
999994  120.009               4            0.699599           0.392061     0  
999999  164.037               4            0.552977           0.454869  2004  

[581909 rows x 11 columns]

In [51]:
y

0         1
1         0
2         1
12        1
13        1
         ..
999990    1
999991    1
999992    1
999994    0
999999    1
Name: song_popularity, Length: 581909, dtype: int64

In [52]:
from sklearn.model_selection import train_test_split

In [53]:
# shift+tab and simply copy
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [54]:
# importing decision tree classifier
from sklearn.tree import DecisionTreeClassifier

In [55]:
#Creating instance "dtree" of the classifier 
dtree = DecisionTreeClassifier(criterion='entropy')

In [56]:
#fitting to the training data, the default parameters are fine at the moment!
dtree.fit(X_train,y_train)

DecisionTreeClassifier(criterion='entropy')

In [57]:
# doing predictions 
dtree_pred = dtree.predict(X_test)

In [58]:
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix

print(classification_report(y_test, dtree_pred))
print(confusion_matrix(y_test, dtree_pred))

              precision    recall  f1-score   support

           0       0.69      0.69      0.69     89287
           1       0.73      0.73      0.73    102743

    accuracy                           0.71    192030
   macro avg       0.71      0.71      0.71    192030
weighted avg       0.71      0.71      0.71    192030

[[61214 28073]
 [27991 74752]]


## Matrix for 10 Features



precision    recall  f1-score   support

           0       0.66      0.66      0.66     89287
           1       0.70      0.70      0.70    102743

    accuracy                           0.68    192030
   macro avg       0.68      0.68      0.68    192030
weighted avg       0.68      0.68      0.68    192030

[[58840 30447]
 [30905 71838]]

In [60]:
# connection.close()